dev/PagerControl/PagerControl.cpp

```diff
@@ -0,0 +1,773 @@
+    {
+        SelectedPageIndex(numberOfPages - 1);
+    }
+    UpdateTemplateSettingElementLists();
+    if (DisplayMode() == winrt::PagerControlDisplayMode::Auto)
+    {
+        UpdateDisplayModeAutoState();
+    }
+    if (numberOfPages > -1)
+    {
+        winrt::VisualStateManager::GoToState(*this, c_finiteItemsModeState, false);
+        if (const auto numberBox = m_numberBox.get())
+        {
+            numberBox.Maximum(numberOfPages);
+        }
+    }
+    else
+    {
+        winrt::VisualStateManager::GoToState(*this, c_infiniteItemsModeState, false);
+        if (const auto numberBox = m_numberBox.get())
+        {
+            numberBox.Maximum(INFINITY);
+        }
+    }
+    UpdateOnEdgeButtonVisualStates();
+}
+
+void PagerControl::OnSelectedPageIndexChange(const int oldValue)
+{
+    // If we don't have any pages, there is nothing we should do.
+    // Ensure that SelectedPageIndex will end up in the valid range of values
+    if (SelectedPageIndex() > NumberOfPages() - 1 && NumberOfPages() > -1)
+    {
+        SelectedPageIndex(NumberOfPages() - 1);
+    }
+    else if (SelectedPageIndex() < 0)
+    {
+        SelectedPageIndex(0);
+    }
+    // Now handle the value changes
+    m_lastSelectedPageIndex = oldValue;
+
+    if (const auto comboBox = m_comboBox.get())
+    {
+        if (SelectedPageIndex() < (int32_t)m_comboBoxEntries.Size())
+        {
+            comboBox.SelectedIndex(SelectedPageIndex());
+        }
+    }
+    if (const auto numBox = m_numberBox.get())
+    {
+        numBox.Value(SelectedPageIndex() + 1);
+    }
+
+    UpdateOnEdgeButtonVisualStates();
+    UpdateTemplateSettingElementLists();
+
+    if (DisplayMode() == winrt::PagerControlDisplayMode::ButtonPanel)
+    {
+        // NumberPanel needs to update based on multiple parameters.
+        // SelectedPageIndex is one of them, so let's do that now!
+        UpdateNumberPanel(NumberOfPages());
+    }
+
+    // Fire value property change for UIA
+    if (const auto peer = winrt::FrameworkElementAutomationPeer::FromElement(*this).try_as<winrt::PagerControlAutomationPeer>())
+    {
+        winrt::get_self<PagerControlAutomationPeer>(peer)->RaiseSelectionChanged(m_lastSelectedPageIndex, SelectedPageIndex());
+    }
+    RaiseSelectedIndexChanged();
+}
+
+void PagerControl::RaiseSelectedIndexChanged()
+{
+    const auto args = winrt::make_self<PagerControlSelectedIndexChangedEventArgs>(m_lastSelectedPageIndex, SelectedPageIndex());
+    m_selectedIndexChangedEventSource(*this, *args);
+}
+
+void PagerControl::OnButtonVisibilityChanged(const winrt::PagerControlButtonVisibility visibility,
+    const wstring_view visibleStateName,
+    const wstring_view collapsedStateName,
+    const wstring_view hiddenStateName,
+    const int hiddenOnEdgePageCriteria)
+{
+    if (visibility == winrt::PagerControlButtonVisibility::Visible)
+    {
+        winrt::VisualStateManager::GoToState(*this, visibleStateName, false);
```

Comment:
Each button (first previous next last) all have independent visibility states right? how does this handle that?

Reply:
Whenever a buttons visibility changes, we pass:
- The new visibility mode
- The name for the buttons visible state
- The name for the buttons collapsed state
- The name for the buttons hidden state
- The selectedindex value that would trigger the "hidden on edge" criteri

When ones button visibility changes, we just call this function with the correct params.